## Libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans


import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Lets download the Geospatial data for Toronto Neighbourhoods

In [2]:
neighb = pd.read_csv('geospat.csv')
neighb.head()

Postcode      Borough                         Neighbourhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Let's see how many neighborhoods and boroughs this dataset has


In [3]:
print('The dataset has {} number of Boroughs and {} number of Neighborhods'.format(len(neighb['Borough'].unique()),len(neighb['Neighbourhood'])))

The dataset has 11 number of Boroughs and 103 number of Neighborhods


### Lets get the latitude and longtude of Toronto City to create a map with its neighbourhoods

In [4]:
address = 'Toronto, Ontario'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

C:\Users\Akshay\Anaconda3\lib\site-packages\geopy\geocoders\osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geograpical coordinate of Toronto City are 43.653963, -79.387207.


In [5]:
# lets create a map of Toronto city using the Latitude and Longitude information  
Toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# Using the location information from geospat file, we add markers on the map for each neighbourhood

for lat, lng, borough, neighborhood in zip(neighb['Latitude'], neighb['Longitude'], neighb['Borough'], neighb['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough) # This will label the markers by neighborhood, borough name
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(Toronto_map)  
    
Toronto_map

## Toronto Neighbourhood Analysis

In our analysis, we will choose "Downtown Toronto" as the borough of interest and then cluster the neighborhoods around it including those in the boroughs with 'Toronto' in their names, e.g. East Toronto, Central Toronto. 

Therefore, int the code below, we will sort the Boroughs that contain the word 'Toronto'.


In [6]:
toronto_data = neighb[neighb['Borough'].str.contains('Toronto')].reset_index(drop = True)
#toronto_data

Postcode           Borough  \
0       M4E      East Toronto   
1       M4K      East Toronto   
2       M4L      East Toronto   
3       M4M      East Toronto   
4       M4N   Central Toronto   
5       M4P   Central Toronto   
6       M4R   Central Toronto   
7       M4S   Central Toronto   
8       M4T   Central Toronto   
9       M4V   Central Toronto   
10      M4W  Downtown Toronto   
11      M4X  Downtown Toronto   
12      M4Y  Downtown Toronto   
13      M5A  Downtown Toronto   
14      M5B  Downtown Toronto   
15      M5C  Downtown Toronto   
16      M5E  Downtown Toronto   
17      M5G  Downtown Toronto   
18      M5H  Downtown Toronto   
19      M5J  Downtown Toronto   
20      M5K  Downtown Toronto   
21      M5L  Downtown Toronto   
22      M5N   Central Toronto   
23      M5P   Central Toronto   
24      M5R   Central Toronto   
25      M5S  Downtown Toronto   
26      M5T  Downtown Toronto   
27      M5V  Downtown Toronto   
28      M5W  Downtown Toronto   
29      M5X  Downtown Toronto   
30      M6G  Downtown Toronto   
31      M6H      West Toronto   
32      M6J      West Toronto   
33      M6K      West Toronto   
34      M6P      West Toronto   
35      M6R      West Toronto   
36      M6S      West Toronto   
37      M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         The Danforth West,Riverdale  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Moore Park,Summerhill East  43.689574 -79.383160  
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         Cabbagetown,St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                           Harbourfront,Regent Park  43.654260 -79.360636  
14                            Ryerson,Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                             Adelaide,King,Richmond  43.650571 -79.384568  
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
25                      Harbord,University of Toronto  43.662696 -79.400049  
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29              First Canadian Place,Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                        Dovercourt Village,Dufferin  43.669005 -79.442259  
32                            Little Portugal,Trinity  43.647927 -79.419750  
33         Brockton,Exhibition Place,Parkdale Village  43.63684

### Get Co-ordinates of Downtown Toronto

In [7]:
address = 'Downtown Toronto, TO'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

C:\Users\Akshay\Anaconda3\lib\site-packages\geopy\geocoders\osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


The geograpical coordinate of Downtown Toronto are 43.655115, -79.380219.


In [8]:
print(*zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']))

(43.67635739999999, -79.2930312, 'The Beaches') (43.6795571, -79.352188, 'The Danforth West,Riverdale') (43.6689985, -79.31557159999998, 'The Beaches West,India Bazaar') (43.6595255, -79.340923, 'Studio District') (43.7280205, -79.3887901, 'Lawrence Park') (43.7127511, -79.3901975, 'Davisville North') (43.7153834, -79.40567840000001, 'North Toronto West') (43.7043244, -79.3887901, 'Davisville') (43.6895743, -79.38315990000002, 'Moore Park,Summerhill East') (43.68641229999999, -79.4000493, 'Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West') (43.6795626, -79.37752940000001, 'Rosedale') (43.667967, -79.3676753, 'Cabbagetown,St. James Town') (43.6658599, -79.38315990000002, 'Church and Wellesley') (43.6542599, -79.3606359, 'Harbourfront,Regent Park') (43.6571618, -79.37893709999999, 'Ryerson,Garden District') (43.6514939, -79.3754179, 'St. James Town') (43.6447708, -79.3733064, 'Berczy Park') (43.6579524, -79.3873826, 'Central Bay Street') (43.65057120000001, -79.3845675, 'Ade

In [9]:
# Lets create a map of Downtown Toronto and its surrounding Neighborhoods
map_dwtor = folium.Map(location=[latitude, longitude], zoom_start=11)

# Showing Downt. Toronto as Red on the map ....
folium.CircleMarker([latitude, longitude], radius=10, popup=label, color='Red', fill=True, fill_color='Red', fill_opacity=0.9,).add_to(map_dwtor)

# ... and the Neighbourhoods as Blue.
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_dwtor)  
    
map_dwtor

## Using Foursquare API

I am going to censor my credentials

In [ ]:
#@hidden_cell
CLIENT_ID = '$$$$'      
CLIENT_SECRET = '$$$$'   
VERSION = '20180901' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

### Lets now explore all the neighborhoods around Downtown Toronto.

The FourSuare API needs latitude and longitudes of the neghborhoods and not the names. Hence, if some neighborhoods have multiple names appended together as one name, won't be a problem for us.

Below we create a function that 
- Extracts name, latitude and longitude of a neighbourhood from our dataset,
- Inputs the client credentials, name, latitude, longitude to the Foursquare API to explore 
  venues within 500m radius of the neighbourhood,
- Retreives Venue data from the Json code that Foursquare API outputs,
- and finally creates a dataframe of all the relevant information we will require to analyze the venues.



In [11]:
def NearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [38]:
radius = 500
LIMIT = 100 # this will retreive data for only 100 venues per neighbourhood
Toronto_venues = NearbyVenues(names = toronto_data['Neighbourhood'], latitudes = toronto_data['Latitude'], longitudes = toronto_data['Longitude'])


The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvall

### Let's take a look at the Venues dataset

In [37]:
print(Toronto_venues.shape)
Toronto_venues.head(10)

(1691, 7)


Neighbourhood  Neighbourhood Latitude  \
0                  The Beaches               43.676357   
1                  The Beaches               43.676357   
2                  The Beaches               43.676357   
3  The Danforth West,Riverdale               43.679557   
4  The Danforth West,Riverdale               43.679557   
5  The Danforth West,Riverdale               43.679557   
6  The Danforth West,Riverdale               43.679557   
7  The Danforth West,Riverdale               43.679557   
8  The Danforth West,Riverdale               43.679557   
9  The Danforth West,Riverdale               43.679557   

   Neighbourhood Longitude                        Venue  Venue Latitude  \
0               -79.293031                    Starbucks       43.678798   
1               -79.293031          Grover Pub and Grub       43.679181   
2               -79.293031                Upper Beaches       43.680563   
3               -79.352188                     Pantheon       43.677621   
4               -79.352188                 Dolce Gelato       43.677773   
5               -79.352188                MenEssentials       43.677820   
6               -79.352188      Messini Authentic Gyros       43.677827   
7               -79.352188              Cafe Fiorentina       43.677743   
8               -79.352188                        Mezes       43.677962   
9               -79.352188  Christina's On The Danforth       43.678240   

   Venue Longitude      Venue Category  
0       -79.298045         Coffee Shop  
1       -79.297215                 Pub  
2       -79.292869        Neighborhood  
3       -79.351434    Greek Restaurant  
4       -79.351187      Ice Cream Shop  
5       -79.351265      Cosmetics Shop  
6       -79.350569    Greek Restaurant  
7       -79.350115  Italian Restaurant  
8       -79.350196    Greek Restaurant  
9       -79.349185    Greek Restaurant

So we can see that there are total 1691 venues in the neighbourhoods in our dataset. The dataframe generated gives interesting information about Venue names like Starbucks or Pantheon and their corresponding categories like Coffee Shop or 
Greek Restaurant. 

We can use this interesting information to categorize these neigborhoods using k_Means clustering to get insights about business or real estate investments.

Lets see number of venues returned per neighbourhood

In [15]:
#pd.set_option('display.max_colwidth', 1)
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              21   
Business reply mail Processing Centre969 Eastern                        16   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      14   
Cabbagetown,St. James Town                                              44   
Central Bay Street                                                      81   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    91   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              36   
Davisville North                                                         7   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      14   
Design Exchange,Toronto Dominion Centre                                100   
Dovercourt Village,Dufferin                                             18   
First Canadian Place,Underground city                                  100   
Forest Hill North,Forest Hill West                                       4   
Harbord,University of Toronto                                           36   
Harbourfront East,Toronto Islands,Union Station                        100   
Harbourfront,Regent Park                                                49   
High Park,The Junction South                                            25   
Lawrence Park                                                            4   
Little Portugal,Trinity                                                 65   
Moore Park,Summerhill East                                               1   
North Toronto West                                                      18   
Parkdale,Roncesvalles                                                   15   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede,Swansea                                                       35   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         92   
Studio District                                                         40   
The Annex,North Midtown,Yorkville                                       24   
The Beaches                                                              3   
The Beaches West,India Bazaar                                           21   
The Danforth West,Riverdale                                             42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Berczy Park                                                              55   
Brockton,Exhibition Place,Parkdale Village                               21   
Business reply mail Processing Centre969 Eastern                         16   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       14   
Cabbagetown,St. James Town                                               44   
Central Bay Street                                                       81   
Chinatown,Grange Park,Kensington Market                                 100   
Christie                                                                 15   
Church and Wellesley                                          

### Venue categories and their densities

Lets now see how many unique venue categories are there and how rare or frequent is their occurence.

In [16]:
Toronto_venues.groupby('Venue Category').count()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                           
Accessories Store                            2                       2   
Adult Boutique                               1                       1   
Afghan Restaurant                            1                       1   
Airport                                      1                       1   
Airport Food Court                           1                       1   
Airport Gate                                 1                       1   
Airport Lounge                               2                       2   
Airport Service                              2                       2   
Airport Terminal                             2                       2   
American Restaurant                         29                      29   
Antique Shop                                 3                       3   
Aquarium                                     4                       4   
Arepa Restaurant                             1                       1   
Art Gallery                                 12                      12   
Art Museum                                   2                       2   
Arts & Crafts Store                          4                       4   
Asian Restaurant                             6                       6   
Auto Workshop                                1                       1   
BBQ Joint                                    4                       4   
Baby Store                                   1                       1   
Bagel Shop                                   5                       5   
Bakery                                      36                      36   
Bank                                         6                       6   
Bar                                         40                      40   
Baseball Stadium                             1                       1   
Basketball Stadium                           4                       4   
Beach                                        1                       1   
Beer Bar                                    11                      11   
Beer Store                                   3                       3   
Belgian Restaurant                           3                       3   
Bistro                                       5                       5   
Boat or Ferry                                1                       1   
Bookstore                                   16                      16   
Boutique                                     1                       1   
Brazilian Restaurant                         2                       2   
Breakfast Spot                              17                      17   
Brewery                                     11                      11   
Bubble Tea Shop                              9                       9   
Building                                     1                       1   
Burger Joint                                26                      26   
Burrito Place                                9                       9   
Bus Line                                     1                       1   
Butcher                                      1                       1   
Café                                        96                      96   
Cajun / Creole Restaurant                    1                       1   
Caribbean Restaurant                         6                       6   
Cheese Shop                                  7                       7   
Chinese Restaurant                          13                      13   
Chocolate Shop                               1                       1   
Church                                       3                       3   
Climbing Gym                                 1                       1   
Clothing Store                              20                      20   
Cocktail Bar                         

We can see from above that Coffe Shop, Caffe, Pubs , Hotels, Italian, Japanese and American Restaurant are the most frequently observed venues.
The occurence of any of these venues in any neighbourhood will define the category of that neighborhood. Hence we will cluster the neighborhoods on the basis of the frequency of the venue categories.

Lets find out how many unique categories we have.

In [17]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


## We now need to find the frequency of occurence of each category in a neighborhood

For that, we first need to convert the venue categories in one-hot vector form such that if a venue category occurs in a particular neighborhood, we assign 1 to that cell(intersection of neighborhood row and venue category column) and 0 for the cells where it doesn't occur.

In [18]:
# one hot encoding
one_hot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")
#print(one_hot.head(10))

In [19]:
# add neighborhood column back to dataframe
one_hot['Neighbourhood'] = Toronto_venues['Neighbourhood'] ## I labeled the new columns as Neighbourhood with a
# 'u' bcs there is already a column named 'Neighborhood' in the venue categories.

# move neighborhood column to the first column
fixed_columns = [one_hot.columns[-1]] + list(one_hot.columns[:-1])
one_hot = one_hot[fixed_columns] 

one_hot.head(5)

Neighbourhood  Accessories Store  Adult Boutique  \
0                  The Beaches                  0               0   
1                  The Beaches                  0               0   
2                  The Beaches                  0               0   
3  The Danforth West,Riverdale                  0               0   
4  The Danforth West,Riverdale                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0             0         0                 0            0           0   
1             0         0                 0            0           0   
2             0         0                 0            0           0   
3             0         0                 0            0           0   
4             0         0                 0            0           0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  \
0                    0                 0              0          0   
1                    0                 0              0          0   
2                    0                 0              0          0   
3                    0                 0              0          0   
4                    0                 0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0 

In [20]:
one_hot.shape

(1691, 232)

### Now lets find the mean frequency of occurence of each category in a neighborhood 

In [47]:
torronto_grouped = one_hot.groupby('Neighbourhood').mean().reset_index()
torronto_grouped.head()

Neighbourhood  Accessories Store  \
0                             Adelaide,King,Richmond               0.01   
1                                        Berczy Park               0.00   
2         Brockton,Exhibition Place,Parkdale Village               0.00   
3   Business reply mail Processing Centre969 Eastern               0.00   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...               0.00   

   Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0             0.0                0.0  0.000000            0.000000   
1             0.0                0.0  0.000000            0.000000   
2             0.0                0.0  0.000000            0.000000   
3             0.0                0.0  0.000000            0.000000   
4             0.0                0.0  0.071429            0.071429   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0      0.000000        0.000000         0.000000          0.000000   
1      0.000000        0.000000         0.000000          0.000000   
2      0.000000        0.000000         0.000000          0.000000   
3      0.000000        0.000000         0.000000          0.000000   
4      0.071429        0.142857         0.142857          0.142857   

   American Restaurant  Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  \
0                 0.04           0.0       0.0               0.0     0.010000   
1                 0.00           0.0       0.0               0.0     0.018182   
2                 0.00           0.0       0.0               0.0     0.000000   
3                 0.00           0.0       0.0               0.0     0.000000   
4                 0.00           0.0       0.0               0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0        0.01                  0.0              0.02         0.0000   
1        0.00                  0.0              0.00         0.0000   
2        0.00                  0.0              0.00         0.0000   
3        0.00                  0.0              0.00         0.0625   
4        0.00                  0.0              0.00         0.0000   

   BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0        0.0         0.0    0.000000  0.020000   0.0  0.030000   
1        0.0         0.0    0.018182  0.036364   0.0  0.000000   
2        0.0         0.0    0.000000  0.000000   0.0  0.047619   
3        0.0         0.0    0.000000  0.000000   0.0  0.000000   
4        0.0         0.0    0.000000  0.000000   0.0  0.000000   

   Baseball Stadium  Basketball Stadium     Beach  Beer Bar  Beer Store  \
0               0.0            0.000000  0.000000  0.000000         0.0   
1               0.0            0.018182  0.018182  0.036364         0.0   
2               0.0            0.000000  0.000000  0.000000         0.0   
3               0.0            0.000000  0.000000  0.000000         0.0   
4               0.0            0.000000  0.000000  0.000000         0.0   

   Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  Boutique  \
0            0.000000  0.000000       0.000000       0.01  0.000000   
1            0.018182  0.018182       0.000000       0.00  0.000000   
2            0.000000  0.000000       0.000000       0.00  0.000000   
3            0.000000  0.000000       0.000000       0.00  0.000000   
4            0.000000  0.000000       0.071429       0.00  0.071429   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                  0.01        0.020000   0.0000              0.0       0.0   
1                  0.00        0.000000   0.0000              0.0       0.0   
2                  0.00        0.095238   0.0000              0.0       0.0   
3                  0.00        0.000000   0.0625              0.0       0.0   
4                  0.00        0.000000   0.0000              0.0       0.0   

   Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0      0.020000    

In [23]:
torronto_grouped.shape ### We have 38 unique neighbourhoods

(38, 232)

## Let's print each neighbourhood with top 5 most frequent of its venues.

In [25]:
num_top_venues = 5

for hood in torronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = torronto_grouped[torronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.07
1                 Café  0.06
2           Steakhouse  0.04
3      Thai Restaurant  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2  Seafood Restaurant  0.04
3          Steakhouse  0.04
4                Café  0.04


----Brockton,Exhibition Place,Parkdale Village----
                  venue  freq
0           Coffee Shop  0.14
1        Breakfast Spot  0.10
2                  Café  0.10
3                   Gym  0.05
4  Caribbean Restaurant  0.05


----Business reply mail Processing Centre969 Eastern----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2        Comic Shop  0.06
3              Park  0.06
4  Recording Studio  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge 

### Lets now find top 10 most common venues per neighborhood and put them into a dataframe

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# Labelling columns as 1st, 2nd and so on
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind])) # for 1st, 2nd, 3rd
    except:
        columns.append('{}th Most Common Venue'.format(ind+1)) ### for 4th, 5th,...

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns) ## assign column names we just created to a new dataframe
neighborhoods_venues_sorted['Neighbourhood'] = torronto_grouped['Neighbourhood']## add neighborhoods column

for ind in np.arange(torronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(torronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3   Business reply mail Processing Centre969 Eastern           Yoga Studio   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse   American Restaurant   
1          Cocktail Bar                Bakery              Beer Bar   
2                  Café        Breakfast Spot  Gym / Fitness Center   
3         Auto Workshop         Garden Center                Garden   
4      Airport Terminal       Airport Service                 Plane   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Thai Restaurant                 Hotel                   Bar   
1              Steakhouse           Cheese Shop    Seafood Restaurant   
2  Furniture / Home Store     Convenience Store             Pet Store   
3      Light Rail Station  Fast Food Restaurant        Farmers Market   
4        Sculpture Garden              Boutique         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Cosmetics Shop            Restaurant                    Gym  
1                  Café        Farmers Market             Restaurant  
2                   Gym          Climbing Gym   Caribbean Restaurant  
3                  Park            Comic Shop       Recording Studio  
4       Harbor / Marina          Airport Gate                Airport

## Clustering the Neighborhoods

In [28]:
# set number of clusters
k = 5

torronto_grouped_clustering = torronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(torronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:38] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 3, 2, 0, 0, 0, 0, 0, 0, 4, 0, 0])

Lets put all data together including the cluster labels, neighborhoods and top 10 venues in each neighborhood and venue name

In [30]:
torronto_merged = toronto_data

# add clustering labels
torronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
torronto_merged = torronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

torronto_merged # check the last columns!

Postcode           Borough  \
0       M4E      East Toronto   
1       M4K      East Toronto   
2       M4L      East Toronto   
3       M4M      East Toronto   
4       M4N   Central Toronto   
5       M4P   Central Toronto   
6       M4R   Central Toronto   
7       M4S   Central Toronto   
8       M4T   Central Toronto   
9       M4V   Central Toronto   
10      M4W  Downtown Toronto   
11      M4X  Downtown Toronto   
12      M4Y  Downtown Toronto   
13      M5A  Downtown Toronto   
14      M5B  Downtown Toronto   
15      M5C  Downtown Toronto   
16      M5E  Downtown Toronto   
17      M5G  Downtown Toronto   
18      M5H  Downtown Toronto   
19      M5J  Downtown Toronto   
20      M5K  Downtown Toronto   
21      M5L  Downtown Toronto   
22      M5N   Central Toronto   
23      M5P   Central Toronto   
24      M5R   Central Toronto   
25      M5S  Downtown Toronto   
26      M5T  Downtown Toronto   
27      M5V  Downtown Toronto   
28      M5W  Downtown Toronto   
29      M5X  Downtown Toronto   
30      M6G  Downtown Toronto   
31      M6H      West Toronto   
32      M6J      West Toronto   
33      M6K      West Toronto   
34      M6P      West Toronto   
35      M6R      West Toronto   
36      M6S      West Toronto   
37      M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
2                       The Beaches West,India Bazaar  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                                  North Toronto West  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                          Moore Park,Summerhill East  43.689574 -79.383160   
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                           Harbourfront,Regent Park  43.654260 -79.360636   
14                            Ryerson,Garden District  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307   
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678   
25                      Harbord,University of Toronto  43.662696 -79.400049   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29              First Canadian Place,Underground city  43.648429 -79.382280   
30                                           Christie  43.669542 -79.422564   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
32                            Little Portugal,Trinity  43.647927 -79.419750   
33         Brockton,Exhibitio

## Visualize the clusters

Lets now visualize the clusters using Folium library.

In [33]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i+x+(i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(torronto_merged['Latitude'], torronto_merged['Longitude'], torronto_merged['Neighbourhood'], torronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining clusters

In [39]:
# Cluster 0, Red

torronto_merged.loc[torronto_merged['Cluster Labels'] == 0, torronto_merged.columns[[1] + list(range(5, torronto_merged.shape[1]))]] #red

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0          Neighborhood   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0                  Café   
4    Central Toronto               0              Bus Line   
5    Central Toronto               0        Sandwich Place   
6    Central Toronto               0           Coffee Shop   
7    Central Toronto               0          Dessert Shop   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0   Japanese Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
22   Central Toronto               0                Garden   
23   Central Toronto               0                  Park   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0                Bakery   
32      West Toronto               0                   Bar   
33      West Toronto               0           Coffee Shop   
34      West Toronto               0       Thai Restaurant   
36      West Toronto               0           Coffee Shop   
37      East Toronto               0           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                     Coffee Shop                   Pub           Yoga Studio   
1                     Coffee Shop        Ice Cream Shop    Italian Restaurant   
2                  Sandwich Place         Movie Theater          Liquor Store   
3                     Coffee Shop    Italian Restaurant             Gastropub   
4                            Park           Swim School    Dim Sum Restaurant   
5                            Park          Burger Joint        Breakfast Spot   
6             Sporting Goods Shop        Clothing Store  Gym / Fitness Center   
7                  Sandwich Place                  Café      Sushi Restaurant   
8                           Diner  Fast Food Restaurant        Farmers Market   
9                             Pub           Pizza Place           Supermarket   
10                     Playground                 Trail           Yoga Studio   
11                     Restaurant     Indian Restaurant    Italian Restaurant   
12                    Coffee Shop               Gay Bar          Burger Joint   
13                           Park                Bakery                  Café   
14                 Clothing Store        Cosmetics Shop                  Café   
15                           Café            Restaurant        Clothing Store   
16                   Cocktail Bar                Bakery              Beer Bar   
18                           Café            Steakhouse   American Restaurant   
19                          Hotel                  Café           Pizza Place   
20                          Hotel                  Café   American Restaurant   
21                          Hotel                  Café            Restaurant   
22                    Yoga Studio  Fast Food Restaurant        Farmers Market   
23                          Trail        

As we can see that most top Venues for cluster label 0 are either Coffe Shop, Sandwich place, Restaurants. We also saw from the map that most neighbourhoods are colored red. So we can conclude that Downtown Toronto is swarming with venues like caffe, Coffee Shop and multicultural restaurants.

Let's examine other clusters

In [42]:
# Cluster 1
torronto_merged.loc[torronto_merged['Cluster Labels'] == 1, torronto_merged.columns[[1] + list(range(5, torronto_merged.shape[1]))]] # purple

Borough  Cluster Labels 1st Most Common Venue  \
24  Central Toronto               1                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24           Coffee Shop        Sandwich Place           Pizza Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24          Burger Joint             BBQ Joint     French Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24     Jewish Restaurant     Indian Restaurant                    Pub

In [43]:
#Cluster 2

torronto_merged.loc[torronto_merged['Cluster Labels'] == 2, torronto_merged.columns[[1] + list(range(5, torronto_merged.shape[1]))]] #blue

Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28                  Café    Seafood Restaurant          Cocktail Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28              Beer Bar                 Hotel            Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28           Cheese Shop                   Pub            Art Gallery

In [45]:
# Cluster 3
# Airport services 
torronto_merged.loc[torronto_merged['Cluster Labels'] == 3, torronto_merged.columns[[1] + list(range(5, torronto_merged.shape[1]))]] #green

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3           Coffee Shop   
27  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café        Sandwich Place    Italian Restaurant   
27      Airport Terminal       Airport Service                 Plane   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                   Bar          Burger Joint   Japanese Restaurant   
27      Sculpture Garden              Boutique         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17       Bubble Tea Shop        Ice Cream Shop     Falafel Restaurant  
27       Harbor / Marina          Airport Gate                Airport

In [46]:
torronto_merged.loc[torronto_merged['Cluster Labels'] == 4, torronto_merged.columns[[1] + list(range(5, torronto_merged.shape[1]))]] #green

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
35  West Toronto               4        Breakfast Spot             Gift Shop   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
35             Piano Bar          Dessert Shop  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
35                   Bar          Burger Joint               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
35         Movie Theater             Restaurant